st3/lsp_utils/npm_client_handler.py

```diff
@@ -0,0 +1,157 @@
+        super().__init__()
+        assert self.package_name
+        self.settings_filename = '{}.sublime-settings'.format(self.package_name)
+        # Calling setup() also here as this might run before `plugin_loaded`.
+        # Will be a no-op if already ran.
+        # See https://github.com/sublimelsp/LSP/issues/899
+        self.setup()
+
+    @classmethod
+    def setup(cls) -> None:
+        assert cls.package_name
+        assert cls.server_directory
+        assert cls.server_binary_path
+        if not cls.__server:
+            cls.__server = ServerNpmResource(cls.package_name, cls.server_directory, cls.server_binary_path)
+        cls.__server.setup()
+
+    @classmethod
+    def cleanup(cls) -> None:
+        if cls.__server:
+            cls.__server.cleanup()
+
+    @property
+    def name(self) -> str:
+        return self.package_name.lower()  # type: ignore
+
+    @property
+    def config(self) -> ClientConfig:
+        assert self.__server
+
+        configuration = {
+            'enabled': True,
+            'command': ['node', self.__server.binary_path] + self.get_binary_arguments(),
+        }
+
+        configuration.update(self._read_configuration())
+        self.on_client_configuration_ready(configuration)
+        return read_client_config(self.name, configuration)
+
+    def get_binary_arguments(self):
+        """
+        Returns a list of extra arguments to append when starting server.
+        """
+        return ['--stdio']
+
+    def _read_configuration(self) -> Dict:
+        settings = {}  # type: Dict
+        loaded_settings = sublime.load_settings(self.settings_filename)  # type: Dict
+
+        if loaded_settings:
+            migrated = self._migrate_obsolete_settings(settings)
+            changed = self.on_settings_read(loaded_settings)
+            if migrated or changed:
+                sublime.save_settings(self.settings_filename)
+
+            for key, default in CLIENT_SETTING_KEYS.items():
+                settings[key] = loaded_settings.get(key, default)
+
+        return settings
+
+    def on_settings_read(self, settings: Dict):
+        """
+        Called when package settings were read. Receives a `sublime.Settings` object.
+
+        Can be used to change user settings, migrating them to new schema, for example.
+
+        Return True if settings were modified to save changes to file.
+        """
+        return False
+
+    def _migrate_obsolete_settings(self, settings: Dict):
+        """
+        Migrates setting with a root `client` key to flattened structure.
+
+        Returns True if settings were migrated.
+        """
+        client = settings.get('client')  # type: Dict
+        if client:
+            settings.erase('client')
+            # Migrate old keys
+            for key, value in client.items():
+                settings.set(key, value)
+            return True
+        return False
+
+    def on_client_configuration_ready(self, configuration: Dict):
```

Comment:
Sorry for the late reply.
Small nitpick on `on_client_configuration_ready(...)`.
Should we add type annotation for the return value?

`on_client_configuration_ready(self, configuration: Dict) -> None: `
This type annotation means that we don't need to return the configuration.

`on_client_configuration_ready(self, configuration: Dict) -> Dict: `
This type annotation means that we do need to return the configuration.

Just to be explicit :)

Reply:
Fixed in 6fc936b7991a72b66f78a26bf7bde88e80288bcd